integration_tests/src/main/python/parquet_write_test.py

```diff
@@ -567,3 +568,76 @@ def test_write_empty_data_single_writer(spark_tmp_path):
+        2. Creates a partitioned Hive, with a similar schema to above, but partitioned on
+           the `type` column.
+        3. Populates the partitioned table using the un-partitioned table as input.
+             1. If the partition mode is dynamic, all the partitions are populated at once.
+             2. If the partition mode is static, the partitions are populated explicitly,
+                and sequentially.
+        4. Returns all the rows from 2 of the 3 partitions in the new table.
+    """
+
+    def create_input_table(spark):
+        tmp_input = spark_tmp_table_factory.get()
+        spark.sql("CREATE TABLE " + tmp_input +
+                  " (make STRING, model STRING, year INT, type STRING, comment STRING)" +
+                  " STORED AS PARQUET")
+        spark.sql("INSERT INTO TABLE " + tmp_input + " VALUES " +
+                  "('Ford',   'F-150',       2020, 'ICE',      'Popular' ),"
+                  "('GMC',    'Sierra 1500', 1997, 'ICE',      'Older'),"
+                  "('Chevy',  'D-Max',       2015, 'ICE',      'Isuzu?' ),"
+                  "('Tesla',  'CyberTruck',  2025, 'Electric', 'Fictional'),"
+                  "('Rivian', 'R1T',         2022, 'Electric', 'Heavy'),"
+                  "('Jeep',   'Gladiator',   2024, 'Hybrid',   'Upcoming')")
+        return tmp_input
+
+    def populate_partitions_static(spark):
+        input_table = create_input_table(spark)
+        output_table = spark_tmp_table_factory.get()
+        spark.sql("CREATE TABLE " + output_table +
+                  " (make STRING, model STRING, year INT, comment STRING)"
+                  " PARTITIONED BY (type STRING) STORED AS PARQUET")
+        spark.sql("INSERT INTO TABLE " + output_table + " PARTITION (type='ICE')" +
+                  " SELECT make, model, year, comment FROM " + input_table +
+                  " WHERE type='Electric'")
+        spark.sql("INSERT OVERWRITE TABLE " + output_table + " PARTITION (type='ICE')" +
+                  " SELECT make, model, year, comment FROM " + input_table +
+                  " WHERE type='Hybrid'")
+        return spark.sql("SELECT * FROM " + output_table +
+                         " WHERE type = 'ELECTRIC' or type = 'Hybrid'")
+
+    def populate_partitions_dynamic(spark):
+        input_table = create_input_table(spark)
+        output_table = spark_tmp_table_factory.get()
+        spark.sql("CREATE TABLE " + output_table +
+                  " (make STRING, model STRING, year INT, comment STRING)"
+                  " PARTITIONED BY (type STRING) STORED AS PARQUET")
+        spark.sql("INSERT OVERWRITE TABLE " + output_table +
+                  " SELECT make, model, year, comment, type FROM " + input_table)
+        return spark.sql("SELECT * FROM " + output_table +
+                         " WHERE type = 'ELECTRIC' or type = 'Hybrid'")
+
+    return populate_partitions_static if mode == PartitionWriteMode.Static else populate_partitions_dynamic
```

Comment:
This should check for PartitionWriteMode.Dynamic explicitly and fail if it is neither Static or Dynamic.  Otherwise we have the common pitfall of blindly assuming if it isn't this it _must_ be the other.  This is Python, so there's no compile-time type checking (someone could be passing a string or an int here by accident). Even if there were, it's a footgun for someone adding a new partition write mode in the future.

Reply:
Agreed. Fixed.